spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+		synchronized (this.updateMap) {
+			try {
+				createNode(key, value);
+				return null;
+			}
+			catch (KeeperException.NodeExistsException e) {
+				// so the data actually exists, we can read it
+				try {
+					byte[] bytes = this.client.getData().forPath(getPath(key));
+					return Conversions.bytesToString(bytes, "UTF-8");
+				}
+				catch (Exception exceptionDuringGet) {
+					throw new ZookeeperMetadataStoreException("Exception while creating node with key '" + key + "':", e);
+				}
+			}
+			catch (Exception e) {
+				throw new ZookeeperMetadataStoreException("Error while trying to set '" + key + "':", e);
+			}
+		}
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		synchronized (this.updateMap) {
+			Stat currentStat = new Stat();
+			try {
+				byte[] bytes = this.client.getData().storingStatIn(currentStat).forPath(getPath(key));
+				if (oldValue.equals(Conversions.bytesToString(bytes, "UTF-8"))) {
+					updateNode(key, newValue, currentStat.getVersion());
+				}
+				return true;
+			}
+			catch (KeeperException.NoNodeException e) {
+				// ignore, the node doesn't exist there's nothing to replace
+				return false;
+			}
+			catch (KeeperException.BadVersionException e) {
+				// ignore
+				return false;
+			}
+			catch (Exception e) {
+				throw new ZookeeperMetadataStoreException("Cannot replace value");
+			}
+		}
+	}
+
+	@Override
+	public void addListener(MetadataStoreListener callback) {
+		this.listeners.add(callback);
+	}
+
+	@Override
+	public void removeListener(MetadataStoreListener callback) {
+		this.listeners.remove(callback);
+	}
+
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		synchronized (this.updateMap) {
+			try {
+				Stat currentNode = this.client.checkExists().forPath(getPath(key));
+				if (currentNode == null) {
+					try {
+						createNode(key, value);
+					}
+					catch (KeeperException.NodeExistsException e) {
+						updateNode(key, value, -1);
+					}
+				}
+				else {
+					updateNode(key, value, -1);
+				}
+			}
+			catch (Exception e) {
+				throw new ZookeeperMetadataStoreException("Error while setting value for key '" + key + "':", e);
+			}
+		}
+	}
+
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		synchronized (this.updateMap) {
+			ChildData currentData = this.cache.getCurrentData(getPath(key));
+			if (currentData == null) {
+				if (this.updateMap.containsKey(key)) {
+					// we have saved the value, but the cache hasn't updated yet
+					// if the value had changed via replication, we would have been notified by the listener
+					return this.updateMap.get(key).getValue();
```

Comment:
You're right, there can be - it's not a problem wrt regular metadata store usage (where eventual consistency after notification has arrived is acceptable), but using the updateMap data first would allow for consistency between returned data and listener invocations.